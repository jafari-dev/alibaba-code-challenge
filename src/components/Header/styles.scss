.header {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  padding-block: 8px;
  box-shadow: 0 8px #aaaaaa20;

  h1 {
    font-weight: 800;
    font-size: 18px;
  }

  div {
    cursor: pointer;
    font-size: 16px;
    font-weight: 600;
    padding-block: 4px;

    svg {
      margin-right: 4px;
      wdith: 32px;
      height: 32px;
      max-width: 32px;
      vertical-align: middle;
    }
  }
}
