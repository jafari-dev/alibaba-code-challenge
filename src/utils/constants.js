export const API_URL = "https://restcountries.com/v2/all";

export const REGION_FILTER_OPTIONS = [
  "All",
  "Africa",
  "Americas",
  "Asia",
  "Europe",
  "Oceania",
];

export const ORDER_FILTERS = {
  nameAsc: "Country name (Asc)",
  nameDes: "Country name (Des)",
  populationAsc: "Population (Asc)",
  populationDes: "Population (Des)",
};

export const ORDER_FILTER_OPTIONS = Object.values(ORDER_FILTERS);

export const THEMES = {
  Light: "light-theme",
  Dark: "dark-theme",
};
