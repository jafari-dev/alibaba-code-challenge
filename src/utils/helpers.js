export function areStringsMatched(str, search) {
  let lastIndex = 0;

  for (const char of search.toLowerCase()) {
    const index = str.toLowerCase().indexOf(char, lastIndex);

    if (index < lastIndex || index === -1) return false;
    else lastIndex = index + 1;
  }

  return true;
}
