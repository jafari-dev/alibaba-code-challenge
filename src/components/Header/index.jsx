import "./styles.scss";
import { Sun, Moon } from "../../assets";
import { memo } from 'react';

function Header({isDarkThemeOn, onToggleTheme}) {
  return (
    <div className="header">
      <header>
        <h1>Where in the world?</h1>
        <div data-testid="theme-switcher" onClick={onToggleTheme}>
          {isDarkThemeOn ?
            <Moon data-testid="theme-icon" /> :
            <Sun data-testid="theme-icon" />}
          <span data-testid="theme-title">{isDarkThemeOn ? "Dark Mode" : "Light Mode"}</span>
        </div>
      </header>
    </div>
  )
}

export default memo(Header);
