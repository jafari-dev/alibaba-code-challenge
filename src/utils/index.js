import { ORDER_FILTERS, THEMES } from "./constants";

export function areStringsMatched(str, search) {
  let lastIndex = 0;

  for (const char of search.toLowerCase()) {
    const index = str.toLowerCase().indexOf(char, lastIndex);

    if (index < lastIndex || index === -1) return false;
    else lastIndex = index + 1;
  }

  return true;
}

function orderObjectsByProperty(
  array,
  property,
  options = { isDescending: false }
) {
  if (options.isDescending === false) {
    return array.sort((a, b) => (a[property] > b[property] ? 1 : -1));
  } else {
    return array.sort((a, b) => (a[property] < b[property] ? 1 : -1));
  }
}

export function reorderCountriesByFilter(countries, filter) {
  if (filter === ORDER_FILTERS.nameAsc) {
    return orderObjectsByProperty(countries, "name");
  } else if (filter === ORDER_FILTERS.nameDes) {
    return orderObjectsByProperty(countries, "name", { isDescending: true });
  } else if (filter === ORDER_FILTERS.populationAsc) {
    return orderObjectsByProperty(countries, "population");
  } else if (filter === ORDER_FILTERS.populationDes) {
    return orderObjectsByProperty(countries, "population", {
      isDescending: true,
    });
  } else {
    return countries;
  }
}

