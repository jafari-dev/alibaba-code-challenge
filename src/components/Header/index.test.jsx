/* eslint-disable testing-library/prefer-screen-queries */
import { fireEvent, render } from "@testing-library/react";
import { faker } from "@faker-js/faker";
import Header from ".";
import { THEMES } from "./../../utils/constants";

beforeAll(() => {
  faker.seed(1);
});

describe("Header component", () => {
  test("matches the snapshots.", () => {
    const { container } = render(
      <>
        <Header theme={THEMES.Light} onToggleTheme={jest.fn()} />
        <Header theme={THEMES.Dark} onToggleTheme={jest.fn()} />
      </>
    );

    expect(container).toMatchSnapshot();
  });

  test("calls the provided 'onToggleTheme' prop when the switcher part is clicked.", () => {
    const onToggleTheme = jest.fn();

    const { getByTestId } = render(
      <Header
        theme={faker.helpers.arrayElement([THEMES.Light, THEMES.Dark])}
        onToggleTheme={onToggleTheme}
      />
    );

    expect(onToggleTheme).toBeCalledTimes(0);

    fireEvent.click(getByTestId("theme-switcher"));
    expect(onToggleTheme).toBeCalledTimes(1);
    fireEvent.click(getByTestId("theme-switcher"));
    expect(onToggleTheme).toBeCalledTimes(2);
  });

  test("contains the sun icon and 'Light Mode' text when the provided 'thme' prop 'light-theme'.", () => {
    const { getByTestId } = render(
      <Header theme={THEMES.Light} onToggleTheme={jest.fn()} />
    );

    expect(getByTestId("theme-title")).toHaveTextContent("Light Mode");
    expect(getByTestId("theme-icon")).toContainHTML("sun.svg");
  });

  test("contains the moon icon and 'Dark Mode' text when the provided 'thme' prop is 'dark-theme'.", () => {
    const { getByTestId } = render(
      <Header theme={THEMES.Dark} onToggleTheme={jest.fn()} />
    );

    expect(getByTestId("theme-title")).toHaveTextContent("Dark Mode");
    expect(getByTestId("theme-icon")).toContainHTML("moon.svg");
  });
});
