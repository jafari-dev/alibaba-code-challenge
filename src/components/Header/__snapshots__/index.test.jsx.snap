// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component matches the snapshots. 1`] = `
<div>
  <div
    class="header"
  >
    <header>
      <h1>
        Where in the world?
      </h1>
      <div
        data-testid="theme-switcher"
      >
        <svg
          data-testid="theme-icon"
        >
          sun.svg
        </svg>
        <span
          data-testid="theme-title"
        >
          Light Mode
        </span>
      </div>
    </header>
  </div>
  <div
    class="header"
  >
    <header>
      <h1>
        Where in the world?
      </h1>
      <div
        data-testid="theme-switcher"
      >
        <svg
          data-testid="theme-icon"
        >
          moon.svg
        </svg>
        <span
          data-testid="theme-title"
        >
          Dark Mode
        </span>
      </div>
    </header>
  </div>
</div>
`;
